config/confidential.go

```diff
@@ -0,0 +1,132 @@
+
+// ConfidentialValue is a string value with a public and a confidential representation
+type ConfidentialValue struct {
+	public, confidential string
+}
+
+func (c ConfidentialValue) Value() string {
+	return c.confidential
+}
+
+func (c ConfidentialValue) String() string {
+	return c.public
+}
+
+func (c *ConfidentialValue) IsConfidential() bool {
+	return c.public != c.confidential
+}
+
+// hideValue hides the entire value in the public representation
+func (c *ConfidentialValue) hideValue() {
+	if c.IsConfidential() {
+		return
+	}
+	c.public = ConfidentialReplacement
+}
+
+// hideSubmatches hides all occurrences of submatches in the public representation
+func (c *ConfidentialValue) hideSubmatches(pattern *regexp.Regexp) {
+	if c.IsConfidential() {
+		return
+	}
+
+	if matches := pattern.FindStringSubmatchIndex(c.confidential); matches != nil && len(matches) > 2 {
+		c.public = c.confidential
+
+		for i := len(matches) - 2; i > 1; i -= 2 {
+			start := matches[i]
+			end := matches[i+1]
+
+			c.public = c.public[0:start] + ConfidentialReplacement + c.public[end:]
+		}
+	}
+}
+
+func NewConfidentialValue(value string) ConfidentialValue {
+	return ConfidentialValue{public: value, confidential: value}
+}
+
+// confidentialValueDecoder implements parsing config parsing for ConfidentialValue
+func confidentialValueDecoder() func(from reflect.Type, to reflect.Type, data interface{}) (interface{}, error) {
+	confidentialValueType := reflect.TypeOf(ConfidentialValue{})
+
+	return func(from reflect.Type, to reflect.Type, data interface{}) (interface{}, error) {
+		if to != confidentialValueType {
+			return data, nil
+		}
+
+		// Source type may be boolean, numeric or string
+		values, isset := stringifyValue(reflect.ValueOf(data))
+		if len(values) == 0 {
+			if isset {
+				values = []string{"1"}
+			} else {
+				values = []string{"0"}
+			}
+		}
+
+		return NewConfidentialValue(values[0]), nil
+	}
+}
+
+// See https://restic.readthedocs.io/en/latest/030_preparing_a_new_repo.html
+var (
+	urlConfidentialPart = regexp.MustCompile("[:/][^:/@]+?:([^:@]+?)@[^:/@]+?") // user:pass@host
+	urlEnvKeys          = regexp.MustCompile("(?i)^.+(_AUTH|_URL)$")
+	hiddenEnvKeys       = regexp.MustCompile("(?i)^(.+_KEY|.+_TOKEN|.*PASSWORD.*|.*SECRET.*)$")
+)
+
+// ProcessConfidentialValues hides confidential parts inside the specified Profile.
+func ProcessConfidentialValues(profile *Profile) {
+	if profile == nil {
+		return
+	}
+
+	// Handle the repo URL
+	profile.Repository.hideSubmatches(urlConfidentialPart)
+
+	// Handle env variables
+	for name, value := range profile.Environment {
+		if hiddenEnvKeys.MatchString(name) {
+			value.hideValue()
+			profile.Environment[name] = value
+		} else if urlEnvKeys.MatchString(name) {
+			value.hideSubmatches(urlConfidentialPart)
+			profile.Environment[name] = value
+		}
+	}
+}
+
+// GetNonConfidentialValues returns a new list with confidential values being replaced with their public representation
```

Comment:
I was working on something on the side, which I pushed to the `shell-escape` branch if you want to have a look.

Basically, instead of having a `map[string][]string` to build a list of arguments, I'm now using a collection (shell.Args) of arguments (shell.Arg) which contain the information on how the argument should be escaped to the shell.

I'm wondering if it's something that could be merged together (information about confidentiality as well as how to escape it to the command line).

**But** that wouldn't satisfy the *separation of concern* principle very well, would it 🤔 

Although I suppose I could still use a `ConfidentialValue` instead of a `string` inside `shell.Arg`. In which case that would simplify the way we construct the real and public command line (and we would have no need for this method <- which was the point of the comment 😄 )

Reply:
First of all: I'm glad that you're working on proper arg escaping 👍. This bit me once but could be solved with `files-from` (and some pre script that creates a list).

To the comment, I see the point. `GetNonConfidentialValues` is a bit of a hacky way to convert an argument list with possible secret values to a public argument list. It is not the best solution and when arguments are escaped it might even break. 

The method could be converted to operate on shell.Args (as input and output) in which case it would have access to the raw values and could map them to public values.

The better approach would probably be to generate a public and an actual list from the sources but it requires changes to the methods that generate shell.Args.

I also think that concerns must be kept separate, otherwise it gets very dirty. So shell.Args should not need to know about public and confidential values. Effectively we have a public and confidential shell.Args instance, either by updating the existing method or by generating the 2 from the config.

Reply:
The implementation works fine on the current codebase, so I'm going to merge it.
Now I'll try to find a nice way to implement proper argument escaping 😄 

Thanks 👍🏻